velox/exec/TableWriter.cpp

```diff
@@ -134,36 +159,88 @@ RowVectorPtr TableWriter::getOutput() {
     fragmentsVector->set(i, StringView(fragments[i - 1]));
   }
 
-  // Set commitcontext column.
-  // clang-format off
-    auto commitContextJson = folly::toJson(
-      folly::dynamic::object
-          (TableWriteTraits::kLifeSpanContextKey, "TaskWide")
-          (TableWriteTraits::kTaskIdContextKey, connectorQueryCtx_->taskId())
-          (TableWriteTraits::kCommitStrategyContextKey, commitStrategyToString(commitStrategy_))
-          (TableWriteTraits::klastPageContextKey, true));
-  // clang-format on
-
+  // 3. Set commitcontext column.
+  const std::string commitContext = createTableCommitContext(true);
   auto commitContextVector = std::make_shared<ConstantVector<StringView>>(
       pool(),
       numOutputRows,
       false /*isNull*/,
       VARBINARY(),
-      StringView(commitContextJson));
+      StringView(commitContext));
 
   std::vector<VectorPtr> columns = {
       writtenRowsVector, fragmentsVector, commitContextVector};
 
+  // 4. Set null statistics columns.
+  if (aggregation_ != nullptr) {
+    for (int i = TableWriteTraits::kStatsChannel; i < outputType_->size();
+         i++) {
+      columns.push_back(BaseVector::createNullConstant(
+          outputType_->childAt(i), writtenRowsVector->size(), pool()));
+    }
+  }
+
   return std::make_shared<RowVector>(
       pool(), outputType_, nullptr, numOutputRows, columns);
 }
 
+std::string TableWriter::createTableCommitContext(bool lastPage) {
+  // clang-format off
+    return folly::toJson(
+      folly::dynamic::object
+          (TableWriteTraits::kLifeSpanContextKey, "TaskWide")
+          (TableWriteTraits::kTaskIdContextKey, connectorQueryCtx_->taskId())
+          (TableWriteTraits::kCommitStrategyContextKey, commitStrategyToString(commitStrategy_))
+          (TableWriteTraits::klastPageContextKey, lastPage));
+  // clang-format on
+}
+
 void TableWriter::updateWrittenBytes() {
   const auto writtenBytes = dataSink_->getCompletedBytes();
   auto lockedStats = stats_.wlock();
   lockedStats->physicalWrittenBytes = writtenBytes;
 }
 
+// static
+RowVectorPtr TableWriter::createAggregationStatsOutput(
+    RowTypePtr outputType,
+    RowVectorPtr aggregationOutput,
+    StringView tableCommitContext,
+    velox::memory::MemoryPool* pool) {
+  // TODO: record aggregation stats output time.
+  if (aggregationOutput == nullptr || aggregationOutput->childrenSize() == 0) {
```

Comment:
then we need to pass in the whole aggregation operator inside, but in this function, we only need its aggregation output, so it would be better to only pass in aggregationOutput